internal/planner/upsert.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2022 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package planner
+
+import (
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+	"github.com/sourcenetwork/defradb/internal/core"
+	"github.com/sourcenetwork/defradb/internal/planner/mapper"
+)
+
+type upsertNode struct {
+	documentIterator
+	docMapper
+
+	p           *Planner
+	collection  client.Collection
+	filter      *mapper.Filter
+	createInput map[string]any
+	updateInput map[string]any
+	doc         *client.Document
+	isUpserting bool
+	results     planNode
+}
+
+// Next only returns once.
+func (n *upsertNode) Next() (bool, error) {
+	if n.isUpserting {
+		next, err := n.results.Next()
+		if err != nil {
+			return false, err
+		}
+
+		switch next {
+		case true:
+			n.currentValue = n.results.Value()
+			// make sure multiple documents do not match
+			next, err := n.results.Next()
+			if err != nil {
+				return false, err
+			}
+			if next {
+				return false, ErrUpsertMultipleDocuments
```

Comment:
question: is there an inherent problemt with upserting multiple docs?

Reply:
There's nothing stopping us from supporting it in the future. It's there to stop users from doing something unexpected like updating all documents when they just wanted to upsert a single document.